### PR TITLE
Implement a first version of the SPDX generator

### PR DESCRIPTION
**Pull request by enzbang**
This is meant only for generating simple SPDX files in the context of e3-based builds.

For #554

### CODE REVIEW

**Review by Nikokrock — APPROVED** [review]

File: src/e3/spdx.py [hunk 1]
```diff
@@ -0,0 +1,791 @@
...
+
+        add_section("Relationships")
+        for rel in self.relationships:
+            output.append(rel.to_tagvalue())
+        for pkg in self.packages.values():
+            add_section("Package")
+            output += pkg.to_tagvalue()
+        return output
+
+    def to_json(self) -> dict[str, Any]:
```
Comment (Nikokrock) [inline src/e3/spdx.py:775]: missing docstring

Comment (Nikokrock) [reply src/e3/spdx.py:775]: call it to dict and implement a to_json that returns a string ?

File: src/e3/spdx.py [hunk 2]
```diff
@@ -0,0 +1,791 @@
...
+class LicenseListVersion(SPDXEntryStr):
+    """Provide the version of the SPDX License List used.
+
+    See 6.7 License list version field
+    """
+
+    pass
+
+
+class Entity(SPDXEntryStr):
```
Comment (Nikokrock) [inline src/e3/spdx.py:227]: Add docstring

File: src/e3/spdx.py [hunk 3]
```diff
@@ -0,0 +1,791 @@
...
+class Created(SPDXEntryStr):
+    """Identify when the SPDX document was originally created.
+
+    See 6.9 Created field
+    """
+
+    pass
+
+
+class Organization(Entity):
```
Comment (Nikokrock) [inline src/e3/spdx.py:272]: missing docstrings for next 3 classes

File: src/e3/spdx.py [hunk 4]
```diff
@@ -0,0 +1,791 @@
+"""Generate a SPDX file.
+
```
Comment (Nikokrock) [inline src/e3/spdx.py:2]: remove all "pass" once all classes have at least a docstring

Comment (enzbang) [reply src/e3/spdx.py:2]: done

File: src/e3/spdx.py [hunk 5]
```diff
@@ -0,0 +1,577 @@
...
+class InvalidSPDX(Exception):
+    pass
+
+
+class SPDXEntry(metaclass=ABCMeta):
+    """Describe a SPDX Entry."""
+
+    @property
+    def entry_key(self) -> str:
```
[the remark this answers is not in the document]
Comment (enzbang) [reply src/e3/spdx.py:38]: added

File: src/e3/spdx.py [hunk 6]
```diff
@@ -0,0 +1,577 @@
...
+        return self.__class__.__name__
+
+    @abstractmethod
+    def __str__(self) -> str:
+        pass
+
+    def __format__(self, format_spec: str) -> str:
+        return self.__str__()
+
+    def to_tagvalue(self) -> str:
```
[the remark this answers is not in the document]
Comment (enzbang) [reply src/e3/spdx.py:54]: added

File: src/e3/spdx.py (hunk 2 above)
Comment (enzbang) [reply src/e3/spdx.py:218]: added

File: src/e3/spdx.py (hunk 3 above)
Comment (enzbang) [reply src/e3/spdx.py:270]: done for all 3

File: src/e3/spdx.py [hunk 7]
```diff
@@ -0,0 +1,577 @@
...
+
+class Relationship(SPDXEntry):
+    """Provides information about the relationship between two SPDX elements.
+
+    See 11.1 Relationship field
+    """
+
+    def __init__(
+        self, left: SPDXID, right: SPDXID, relationship_type: RelationshipType
+    ) -> None:
```
[the remark this answers is not in the document]
Comment (enzbang) [reply src/e3/spdx.py:518]: updated

File: src/e3/spdx.py [hunk 8]
```diff
@@ -0,0 +1,773 @@
...
+
+
+if TYPE_CHECKING:
+    from typing import Literal, Union, Any
+
+NOASSERTION: Literal["NOASSERTION"] = "NOASSERTION"
+NONE_VALUE: Literal["NONE"] = "NONE"
+
+if TYPE_CHECKING:
+    MAYBE_STR = Union[str, Literal["NOASSERTION"], Literal["NONE"]]
```
Comment (Nikokrock) [inline src/e3/spdx.py:25]: technically str | Literal["somestr"] = str ;-)

**Review by Nikokrock — APPROVED** [review]

**Pull request merged**